### DISCUSSION TITLE
Various issues with SPA feature

### DISCUSSION THREAD

**Discussion by mkarajohn** [remix-run/remix#8534] [Q&A]
Context: I am trying to (experimentally) migrate an existing Vite + RR5 app to Remix SPA and this has turned into an ordeal.

So, I followed the [Migrating from RR](https://remix.run/docs/en/main/guides/migrating-react-router-app) guide and I ended up with the following:

```tsx
// app/routes/_index.tsx

export default function Index() {
  return <h1>hello</h1>;
}
```

```tsx
// app/routes/$.tsx

export default function Index() {
  return <h1>hello</h1>;
}
```

```tsx
// app/entry.client.tsx

import { RemixBrowser } from "@remix-run/react";
import { startTransition, StrictMode } from "react";
import { hydrateRoot } from "react-dom/client";

startTransition(() => {
  hydrateRoot(
    document,
    <StrictMode>
      <RemixBrowser />
    </StrictMode>
  );
});
```

```tsx
// app/entry.server.tsx

import type { EntryContext } from '@remix-run/node';
import { RemixServer } from '@remix-run/react';
import { renderToString } from 'react-dom/server';

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  // responseHeaders: Headers,
  remixContext: EntryContext
) {
  const html = renderToString(<RemixServer context={remixContext} url={request.url} />);

  return new Response(html, {
    headers: { 'Content-Type': 'text/html' },
    status: responseStatusCode,
  });
}
```
```
// app/env.d.ts

/// <reference types="@remix-run/node" />
/// <reference types="vite/client" />
```

```tsx
// app/tsx root.tsx

import { Links, LiveReload, Meta, Outlet, Scripts, ScrollRestoration } from '@remix-run/react';

export default function App() {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        <Outlet />
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}

export function HydrateFallback() {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        <p>yolo...</p>
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}
```

```ts
// vite.config.ts

import { unstable_vitePlugin as remix } from "@remix-run/dev";
import { defineConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";

export default defineConfig({
  plugins: [remix({ unstable_ssr: false }), tsconfigPaths()],
});
```

```json
// tsconfig.json

{
  "include": ["env.d.ts", "**/*.ts", "**/*.tsx"],
  "compilerOptions": {
    "lib": ["DOM", "DOM.Iterable", "ES2022"],
    "isolatedModules": true,
    "esModuleInterop": true,
    "jsx": "react-jsx",
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "resolveJsonModule": true,
    "target": "ES2022",
    "strict": true,
    "allowJs": true,
    "forceConsistentCasingInFileNames": true,
    "baseUrl": ".",
    "paths": {
      "~/*": ["./app/*"]
    },
    "noEmit": true
  }
}
```
Some relevant installed packages:

```json
    "@remix-run/node": "^2.5.0",
    "@remix-run/react": "^2.5.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.21.2",
    "@remix-run/dev": "^2.5.0",
    "typescript": "~5.2.2",
    "vite": "^5.0.5",
    "vite-tsconfig-paths": "^4.2.2",
```

Running `remix vite:dev` would initially throw an error about `RefreshRuntime` being declared twice and then it started throwing the error from the title. 

```
Unexpected Server Error  TypeError: Cannot read properties of undefined (reading 'routes')
```

I have deleted the `index.html` and `main.tsx` that were used in the vanilla Vite configuration. 

I have spent too much time on this, and I am this close to giving up on it 😮‍💨

The full error trace is this:

```sh
TypeError: Cannot read properties of undefined (reading 'routes')
    at RemixServer (/home/mitch/Workspace/github/app-name/node_modules/@remix-run/react/dist/server.js:60:53)
    at renderWithHooks (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:5662:16)
    at renderIndeterminateComponent (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:5735:15)
    at renderElement (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:5950:7)
    at renderNodeDestructiveImpl (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6108:11)
    at renderNodeDestructive (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6080:14)
    at retryTask (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6532:5)
    at performWork (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6580:7)
    at /home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6904:12
    at scheduleWork (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:78:3)
    at startWork (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6903:3)
    at renderToStringImpl (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6977:3)
    at Module.renderToString (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:7065:10)
    at handleRequest (/home/mitch/Workspace/github/app-name/app/entry.server.tsx:11:16)
    at handleDocumentRequestRR (/home/mitch/Workspace/github/app-name/node_modules/@remix-run/server-runtime/dist/server.js:226:18)
    at async requestHandler (/home/mitch/Workspace/github/app-name/node_modules/@remix-run/server-runtime/dist/server.js:105:18)
    at async /home/mitch/Workspace/github/app-name/node_modules/@remix-run/dev/dist/vite/node/adapter.js:84:20
    at async /home/mitch/Workspace/github/app-name/node_modules/@remix-run/dev/dist/vite/plugin.js:598:15
TypeError: Cannot read properties of undefined (reading 'routes')
    at RemixServer (/home/mitch/Workspace/github/app-name/node_modules/@remix-run/react/dist/server.js:60:53)
    at renderWithHooks (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:5662:16)
    at renderIndeterminateComponent (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:5735:15)
    at renderElement (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:5950:7)
    at renderNodeDestructiveImpl (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6108:11)
    at renderNodeDestructive (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6080:14)
    at retryTask (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6532:5)
    at performWork (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6580:7)
    at /home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6904:12
    at scheduleWork (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:78:3)
    at startWork (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6903:3)
    at renderToStringImpl (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6977:3)
    at Module.renderToString (/home/mitch/Workspace/github/app-name/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:7065:10)
    at handleRequest (/home/mitch/Workspace/github/app-name/app/entry.server.tsx:11:16)
    at handleDocumentRequestRR (/home/mitch/Workspace/github/app-name/node_modules/@remix-run/server-runtime/dist/server.js:254:20)
    at async requestHandler (/home/mitch/Workspace/github/app-name/node_modules/@remix-run/server-runtime/dist/server.js:105:18)
    at async /home/mitch/Workspace/github/app-name/node_modules/@remix-run/dev/dist/vite/node/adapter.js:84:20
    at async /home/mitch/Workspace/github/app-name/node_modules/@remix-run/dev/dist/vite/plugin.js:598:15
```

**Comment by brophdawg11** [comment]
I think this is your problem:

```ts
export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  // responseHeaders: Headers,
  remixContext: EntryContext
) {...}
```

You commented out a parameter for the function so `remixContext` is not getting the proper value.  You need to leave the 3rd parameter in there even though you aren't using it for SPA mode's rendering of index.html.

**Reply by brophdawg11** [reply]
The code for `<RemixServer>` is the following, which is why you're getting the `undefined` error on `routes`:

```jsx
export function RemixServer({ context, url, abortDelay }) {
  if (typeof url === "string") {
    url = new URL(url);
  }

  let { manifest, routeModules, criticalCss, serverHandoffString } = context;
  let routes = createServerRoutes(
    manifest.routes, // 💥
    routeModules,
    context.future,
    context.isSpaMode
  );
```

**Reply by mkarajohn** [reply]
Oh my god, I can't believe that, I was reading the arguments as if it was object destructuring instead of discrete arguments :facepalm: ... To my defense 15 hours ago was 3am local time :sweat_smile: 

Ok, so this does fix that problem, but now I am getting this:

```
Uncaught SyntaxError: Identifier 'RefreshRuntime' has already been declared (at root.tsx:2:1)
```

Where the start of the compiled file, as shown in the console is this:
```js
import { createHotContext as __vite__createHotContext } from "/@vite/client";import.meta.hot = __vite__createHotContext("/app/root.tsx");import RefreshRuntime from "/@id/__x00__virtual:remix/hmr-runtime";const inWebWorker = typeof WorkerGlobalScope !== 'undefined' && self instanceof WorkerGlobalScope;let prevRefreshReg;let prevRefreshSig;if (import.meta.hot && !inWebWorker && window.__remixLiveReloadEnabled) {  if (!window.__vite_plugin_react_preamble_installed__) {    throw new Error(      "Remix Vite plugin can't detect preamble. Something is wrong."    );  }  prevRefreshReg = window.$RefreshReg$;  prevRefreshSig = window.$RefreshSig$;  window.$RefreshReg$ = (type, id) => {    RefreshRuntime.register(type, "/home/mitch/Workspace/github/app-name/app/root.tsx" + " " + id)  };  window.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;}import { jsxDEV } from "/node_modules/.vite/deps/@emotion_react_jsx-dev-runtime.js?v=0c7f045f";
import RefreshRuntime from "/@react-refresh"; // <-- second RefreshRuntime gets imported
```

**Reply by brophdawg11** [reply]
Don't worry - you are not alone!  I do that almost every time I go in and edit `entry.server.tsx` 😂 

As for the ReactRefresh issue - it's hard to say just from the above code.  Any chance you can provide a minimal reproduction in stackblitz or a GH repo?

**Reply by mkarajohn** [reply]
Ok, figured this one. 

It occurs if the `react` plugin is used alongside the `remix` plugin. There needs to be only 1 (remix)

While the docs (https://remix.run/docs/en/main/future/vite and https://remix.run/docs/en/main/future/spa-mode) don't ever show using them both, they don't explicitly say you shouldn't use them both. I plan to document the whole process of the migration, but until then maybe this detail should be mentioned in the docs. 

Which brings me to the next question:

Since we no longer use the `react` plugin, how do we apply babel plugins? For example, what's the equivalent for this:

```js
      react({
        babel: {
          plugins: ['@emotion/babel-plugin'],
        },
      }),
```

Should we move to the standalone babel-plugin? https://www.npmjs.com/package/vite-plugin-babel

**Reply by mkarajohn** [reply]
Ok, now I changed the `$.tsx` and `_index.tsx` to actually import the real app (baby steps, yay)

```js
// $.tsx and _index.tsx

export { default } from '../App.tsx';
```

 and I got the css import error (one of our dependencies is doing `import 'tippy.css'`, which vanilla vite was handling fine, but something is not quite the same with the `remix` plugin)

![css_import](https://github.com/remix-run/remix/assets/3789226/3bb41f10-237a-40ea-9e5f-436ecc5512b7)

I had the same error previously and I simply patched the dependency to comment out the css import for now. 

The error that showed up after this was due to (the same) dependency trying to use `navigator` inside an IIFE, for which you told me that everything that gets rendered inside the HydrateFallback needs to be ssr safe, but how do we do that when it's the whole `App.tsx` component that gets rendered? :sweat_smile: 

~~Can I opt-out the App.tsx component from being evaluated on the server?~~ 

~~UPDATE: Got it https://remix.run/docs/en/main/discussion/server-vs-client#forcing-code-out-of-the-browser-or-server-builds~~

UPDATE 2: didn't get it:

```js
// $.tsx and _index.tsx

export { default } from '../App.client.tsx';
```

```
Error: You made a GET request to "/" but did not provide a `loader` for route "routes/_index", so there is no way to handle the request.
```

**Reply by mkarajohn** [reply]
Ok, I reverted `App.client.tsx` back to `App.tsx` and , you guessed right, commented out the problematic parts of the above package, in order to be able to continue and now I got this from another package:

```
[vite] Error when evaluating SSR module virtual:remix/server-build: failed to import "/src/routes/_index.tsx"
|- /home/mitch/Workspace/github/app-name/node_modules/package-name/dist/esm/index.js:1
export * from './contexts';
^^^^^^

SyntaxError: Unexpected token 'export'
```

I read this part of the docs but it did not help https://remix.run/docs/en/main/future/vite#esm--cjs :/

**Reply by brophdawg11** [reply]
> Since we no longer use the react plugin, how do we apply babel plugins

I think there's been talk of potentially letting you pass them through the `remix()` plugin - will check on that.  But for now I'd try the standalone babel plugin and see if that does the trick

> but how do we do that when it's the whole App.tsx component that gets rendered

So do you have a dependency that _on import_ executes an IIFE that tries to reach for navigator?  If so can you use an async import for that dependency and load it from client side code?

> and now I got this from another package:

CJS/ESM is not my strong suit unfortunately.  Would probably need a small repro of an app using the specific dependency to be able to dig in further

**Reply by mkarajohn** [reply]
An update on this: unfortunately I don't have enough time to give on this; I started it as an experiment to see if it would be possible to migrate our Vite + RR project to Remix SPA, expecting I would not hit any major roadblocks. Considering all the aforementioned issues (excluding the commented out argument :sweat_smile: ) this seems to not be ready yet (duh, it's "unstable") for our case. 

I will give it another shot when the feature comes out as stable.

**Reply by mkarajohn** [reply]
btw, this package

```
[vite] Error when evaluating SSR module virtual:remix/server-build: failed to import "/src/routes/_index.tsx"
|- /home/mitch/Workspace/github/app-name/node_modules/package-name/dist/esm/index.js:1
export * from './contexts';
^^^^^^

SyntaxError: Unexpected token 'export'
```

 exports its modules like this, in case this helps:

```
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "types": "dist/types/index.d.ts",
  "exports": {
    "types": "./dist/types/index.d.ts",
    "import": "./dist/esm/index.js",
    "require": "./dist/cjs/index.js",
    "default": "./dist/cjs/index.js"
  },
  ```

**Reply by mkarajohn** [reply]
@brophdawg11 
Ok, a little progress on this.

The package above was missing the  `"type": "module"` field. Once added the error about "unexpected token export" disappeared but a new one appeared:

Now the bundler (always using the remix vite plugin)  cannot figure out imports without a file extension, or in the case of `index` exports, it needs the full path to the file (so for example `import X from '../components/X'` needs to become `import X from '../components/X/index.js'` (it needs them to strictly follow the ESM spec, in short)

I went ahead and changed every single import in the package manually, to comply with the above, but now the next error is that imports of svg files (e.g. `import asset from '../assets/asset.svg'`) cannot be resolved, as expected, since it seems to operate in a strict ESM mode and not a "bundler" mode, which knows how to handle such things.

To me it seems like something is off with the bundler settings when the remix plugin is used. This is not behaviour I have ever faced with vanilla Vite.

Hope this narrows the scope a little so hopefully someone can look into it.

**Reply by brophdawg11** [reply]
Would you be able to provide a reproduction?  

ESM/CJS resolution has a lot to do with configurations outside of Remix so it's hard to triage this specific issue without knowing the full scope of everything.  None of our packages have `"type": "module"` which I believe is by design because they may not run in `"type":"module"` apps - and that's why they have `main`/`module` `exports` fields - so your bundler can choose the right file.

So I have to think that this is something about your local application/bundler configuration picking up the wrong file for some reason.

**Reply by mkarajohn** [reply]
@brophdawg11 

Here you go :pray:  https://stackblitz.com/edit/vitejs-vite-u1xsya?file=vite.config.ts

**Reply by brophdawg11** [reply]
Ok there's a few things going on here:

* I didn't realize you weren't talking about a remix dependency being the CJS/ESM issue.  I am not familiar with `@orfium/toolbox` but that seems to be the issue.  Removing that import gets rid of the ESM/CJS issue.  Have you checked their repo to look for any similar issues of ESM/CJS/Vite compatibility by chance?
* `routes/root.tsx` is empty which is problematic since that is what renders the main HTML page.  This is discussed in [this section of the migration guide](https://remix.run/docs/en/main/guides/migrating-react-router-app#creating-the-root-route) for more information - but your best bet is to grab the [version](https://github.com/remix-run/remix/blob/main/templates/spa/app/root.tsx) from the Remix SPA template and use/adapt that to your needs
* You don't want to be rendering your own `BrowserRouter` in Remix SPA mode since Remix already does that under the hood for you via `<RemixBrowser>` - so you only need to include your `<Routes>` tree.  Here's the relevant [section of the migration guide](https://remix.run/docs/en/main/guides/migrating-react-router-app#adapting-your-existing-app-code) for more info.
* I didn't have any issues importing and renderoing the SVG so unsure what that issue was

Here's a working code sandbox example with those issues resolved: https://stackblitz.com/edit/vitejs-vite-sqgue6

**Reply by mkarajohn** [reply]
Thanks for the quick reponse!

RE: `routes/root.tsx` : my bad, this was an error on my part when I made the stackblitz project, the real one on my machine does have a valid `root.tsx` :muscle: 

RE: `Browserrouter`: :+1: 

RE: `@orfium/toolbox`: I happen to be one of the maintainers for this :P and as far as I know we ("we" actually incudes me as well, since I use it in the product I work on for my org) never had any issues with neither CRA (when we used it), nor Vite (without the remix plugin) nor with Vitest. [This](https://github.com/Orfium/toolbox/blob/12-19-chore_delete_unused_state_in_Authentication_provider/package.json) is the `package.json` of the version included in the repro demo (and my local project), I don't see anything outright wrong with it, but let me know if you see something I don't.. Btw, [these lines](https://github.com/Orfium/toolbox/blob/12-19-chore_delete_unused_state_in_Authentication_provider/package.json#L91-L92) are the build scripts, let me know if you see anything weird there as well. Not using this is non-negotiable since it is required in all the projects of my org. Thanks

**Reply by jrestall** [reply]
Handy tool for analyzing ESM/CJS package issues. https://publint.dev/@orfium/toolbox@1.15.0

**Reply by mkarajohn** [reply]
@jrestall oh thanks, that's a killer tool! I did not know it existed!
Will work on resolving the issues it throws and update the thread. And here I was thinking I had figured out esm/cjs bundling :sweat_smile: 

For the record though, I will note again that Vite, without the remix plugin, has no issue consuming the aforementioned package. Something works differently when the remix plugin is used

**Reply by mkarajohn** [reply]
Update:

Exports are fixed https://publint.dev/@orfium/toolbox@2.0.0-rc.43 :confetti_ball: Really glad I learnt about the publint tool

Now I am getting the issue with the .svg import: https://stackblitz.com/edit/vitejs-vite-u1xsya?file=src%2FApp.tsx
(the @orfium/toolbox package is importing an svg asset in one of its components, the offending import can be found [here](https://github.com/Orfium/toolbox/blob/01-25-feat_convert_the_package_to_proper_ES_module/src/ui/Navigation/components/GlobalNav/GlobalNav.tsx#L5))

I am not getting this error when I run the project with Vite using the following plugins:

```js
plugins: [
      react({
        babel: {
          plugins: ['@emotion/babel-plugin'],
        },
      }),
      tsconfigPaths(),
      svgr(),
      ViteEjsPlugin(),
    ],
```

this error only occurs when I try to run it using `remix vite:dev`

It seems as if it is not properly handling svg imports that are not in user-code directly (as in the [working example](https://stackblitz.com/edit/vitejs-vite-sqgue6) provided by @brophdawg11 )

**Reply by brophdawg11** [reply]
ok - would you mind opening a new github issue with a minimal reproduction of just the SVG issue - and remove all the other unrelated aspects of this setup?  That will allow the team to look at that SVG issue alone 👍

**Reply by mkarajohn** [reply]
[Done](https://github.com/remix-run/remix/issues/8606), thanks for all the support :muscle:

**Reply by JoepKockelkorn** [reply]
Just to add my two cents on SPA mode issues: I ran into [this issue](https://remix.run/docs/en/main/future/vite#server-code-errors-in-browser-during-development), but I didn't fix it by isolating server stuff to `.server.ts` files but rather by replacing any `@remix-run/node` imports by their `@remix-run/react` equivalents. Because `redirect`ing is something you still do in `clientLoader`s.
